src/test/java/com/radixdlt/atommodel/validators/ValidatorConstraintScryptTest.java

```diff
@@ -0,0 +1,216 @@
+package com.radixdlt.atommodel.validators;
+
+import com.radixdlt.atommodel.validators.ValidatorConstraintScrypt.ValidatorTransitionProcedure;
+import com.radixdlt.atomos.CMAtomOS;
+import com.radixdlt.atomos.Result;
+import com.radixdlt.constraintmachine.Particle;
+import com.radixdlt.constraintmachine.Spin;
+import com.radixdlt.constraintmachine.WitnessData;
+import com.radixdlt.identifiers.RadixAddress;
+import com.radixdlt.store.CMStore;
+import com.radixdlt.store.CMStores;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.util.function.Function;
+import java.util.function.UnaryOperator;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Exercise the critical invariants of the {@link ValidatorConstraintScrypt}.
+ *
+ * The invariants tested here follow the flow of the validator FSM:
+ *  - unregistered validator with nonce 0 is virtualised as UP
+ *  - validator address must be specified
+ *  - validator transitions must be between validators of the same address
+ *  - the nonce must increase by exactly one with every transition
+ *  - validator transitions must be signed by the address
+ *
+ * These tests are implemented by either testing the static check directly or the generated transition procedure.
+ */
+public class ValidatorConstraintScryptTest {
+	private static Function<Particle, Result> staticCheck;

```

Comment:
Not sure why these are initialised once for the whole class.  Feels like they should not be static, and the initialisation should be @ Before, not @ BeforeClass.

Reply:
Why does it feel like they shouldn't be static?

Reply:
They're supposed to be the same for all tests

Reply:
I think they should be equivalent for each test, but having them use the exact same objects is potentially problematic, eg having state retained between test methods.

Reply:
Fair enough. Currently, this shouldn't be an issue by design, but if it ever becomes one, this isn't the right place to test it. Will change.